File: src/java/graph/bfs/tests/TraversalTests.java
package graph.bfs.tests;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import graph.bfs.Traversal;


public class TraversalTests {

    private Map<String, List<String>> graph = Map.of(
            // first subgraph
            "A", List.of("B", "C", "D"),
            "B", List.of("E"),
            "C", List.of("F", "G"),
            "F", List.of("H", "I"),

            // second subgraph
            "J", List.of("K", "L", "M"),
            "L", List.of("N", "O")
    );

    @Test                                               
    @DisplayName("successful graph traversal using breadth-first search")   
    public void successful_traversal() {
        var firstSubgraph = Traversal.traverse(graph, "A");
        assertThat(firstSubgraph).containsExactly("A", "B", "C", "D", "E", "F", "G", "H", "I");

        var secondSubgraph = Traversal.traverse(graph, "J");
        assertThat(secondSubgraph).containsExactly("J", "K", "L", "M", "N", "O");
    }

    @Test
    @DisplayName("throws exception if the root vertex is invalid")   
    public void exception_if_invalid_root() {
        assertThatThrownBy(() -> Traversal.traverse(graph, "InvalidVertex"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Invalid root vertex");
    }
}

File: src/kotlin/graph/bfs/tests/TraversalTests.kt
package graph.bfs.tests

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.assertions.throwables.shouldThrowExactly
import graph.bfs.traverse


class TraversalTests : FunSpec({

   val graph = mapOf(
         // first subgraph
         "A" to listOf("B", "C", "D"),
         "B" to listOf("E"),
         "C" to listOf("F", "G"),
         "F" to listOf("H", "I"),

         // second subgraph
         "J" to listOf("K", "L", "M"),
         "L" to listOf("N", "O")
   )

   test("successful graph traversal using breadth-first search") {
      val firstSubgraph = graph.traverse("A")
      firstSubgraph shouldBe listOf("A", "B", "C", "D", "E", "F", "G", "H", "I")

      val secondSubgraph = graph.traverse("J")
      secondSubgraph shouldBe listOf("J", "K", "L", "M", "N", "O")
   }

   test("throws exception if the root vertex is invalid") {
      shouldThrowExactly<IllegalArgumentException> {
         graph.traverse("InvalidVertex")
      }
   }
})